fix(broadcast): ported to 12.x

samuelmasse opened the pull request:
This upgrades broadcast to botpress 12 (or 13)

samuelmasse committed:
basic working state

samuelmasse committed:
delete fix

samuelmasse committed:
working messages

samuelmasse committed:
remove filters

allardy commented on modules/broadcast/src/backend/daemon.ts:
```diff
@@ -17,35 +17,11 @@ export default async (botId: string, bp: SDK, db: Database) => {
-          return typeof v !== 'undefined' && v !== null && v !== true
-        })
-      })
-    }
```
I still think filters are an important part of broadcast. Without them, impossible to ignore users who haven't spoke to the bot in the last month, and also impossible to filter unwanted channels (eg: api when benchmarking creates a ton of users).

allardy commented on modules/broadcast/src/backend/daemon.ts:
```diff
@@ -17,35 +17,11 @@ export default async (botId: string, bp: SDK, db: Database) => {
-
-      const content = await bp.cms.getContentElement(botId, row.text)
+      const botInfo = await bp.bots.getBotById(botId)
+      const content = await bp.cms.getContentElement(botId, row.text, botInfo.defaultLanguage)
```
If a user has configured a language, we might want that one instead of the bot default

allardy commented on modules/broadcast/src/backend/db.ts:
```diff
@@ -191,7 +192,7 @@ export default class BroadcastDb {
         this.where(isPast).andWhere('broadcast_outbox.botId', botId)
       })
-      .join('srv_channel_users', 'srv_channel_users.user_id', 'broadcast_outbox.userId')
+      .join('srv_channel_users', 'srv_channel_users.id', 'broadcast_outbox.userId')
```
I think it would be better to keep user_id, since it's the value we use everywhere to reference a user. The numeric ID has no value outside that table

allardy reviewed (commented):
I think there is still some work to do before we add it as a "non-dotted" module. I also tested it, I received the message, but also got spammed with a ton of "undefined" in the console

- Update the readme 
- Remove unused files (npmignore, outdated changelog)
- Add the tsconfig.json file at the root, and the tsconfig / tslint file to the views folder
- The lock mechanism should use `bp.distributed` so the lock mechanism is handled by Redis and not the KVS

Other than that, the UI could use a small overhaul so it fits with the rest of the modules.

EFF commented:
Please change the target branch to 12.8 as this is a minor fix and doesn't require a major release

samuelmasse committed:
Revert "remove filters"

This reverts commit 706130833be68dadbe19b5203e1db098c227e3ad.

samuelmasse committed:
quickfixes

samuelmasse committed:
fix & user language

samuelmasse committed:
redis locks

samuelmasse committed:
Update README.md

allardy commented:
There's a couple of issues before this PR can be merged:

1. Since this is not officially supported, I wouldn't add information on the documentation page, but the readme should be updated with up-to-date information. There's still information on

2. There's no information on how to use filtering, and the old method cannot be used anymore (not possible to register methods on bp). It's a free text field and even I have no clue on how to use it

3. I'd like to have 1 or two examples (use case) that can be used to test the module and get started

4. The timezone is commented in the code, but the checkbox is still in the UI

samuelmasse committed:
Merge remote-tracking branch 'origin/minor/12-10-0' into sm-broadcast

samuelmasse committed:
Update pro

samuelmasse committed:
Revert "Update pro"

This reverts commit c6096cb2fab9300320d3ee94ad51ff06a40de562.

samuelmasse committed:
fix z-index

samuelmasse committed:
change deprecated

samuelmasse committed:
await

samuelmasse committed:
fix filter

samuelmasse committed:
readme

allardy commented:
The module is a bit chatty, and I don't have much users..

![image](https://user-images.githubusercontent.com/42552874/80829904-00af0400-8bb6-11ea-915f-11ee95b13a7e.png)

samuelmasse committed:
useless files

samuelmasse committed:
port to blueprint

samuelmasse committed:
round 2

samuelmasse committed:
modal buttons

samuelmasse committed:
buttons & icons

samuelmasse committed:
other buttons

samuelmasse committed:
filters

samuelmasse committed:
finish modal

samuelmasse committed:
navbar

samuelmasse committed:
buttons refactor

samuelmasse committed:
style

samuelmasse committed:
remove bootstrap

samuelmasse committed:
remove boostrap deps

samuelmasse committed:
style error

samuelmasse committed:
error placement

samuelmasse committed:
fix flickering

samuelmasse committed:
remove duplicate error

samuelmasse committed:
setup translations

samuelmasse committed:
translations

samuelmasse commented:
@allardy I think this could use a bit of refactoring (converting to .tsx + typings) but for the user side of things I'd say it's at a good enough state. I removed all the bootstrap and replaced it with blueprint, fixed the flickering ui, added translations and reworked the form ui.

allardy commented:
Good job with the blueprint transition! There are some usability issues tho: 

1. If a button can be disabled, it needs to be an AnchorButton, not a Button (otherwise the tooltip doesn't go away).
2. You can create broadcasts in the past, which are run automatically but seems stuck at a weird percentage, and always stay "in progress"

![image](https://user-images.githubusercontent.com/42552874/82404001-277e8e80-9a2e-11ea-9dca-371b3c358d6e.png)

3. Sending content elements actually seems to send empty messages
![image](https://user-images.githubusercontent.com/42552874/82404118-70cede00-9a2e-11ea-9d38-6695c43534e6.png)

4. Maybe put a link or something to some information about filters, because when you create a broadcast, you have absolutely no idea what can be put in the filters section

![image](https://user-images.githubusercontent.com/42552874/82404182-9fe54f80-9a2e-11ea-88fe-2ae8adf4fdce.png)

samuelmasse committed:
anchor buttons

samuelmasse committed:
fix content elements

samuelmasse committed:
docs link

samuelmasse committed:
fix title

samuelmasse committed:
min & max dates

EFF commented:
@samuelmasse can you add this module as experimental then we will be good to ship

allardy committed:
Merge branch 'minor/12-10-0' into sm-broadcast

allardy commented:
I found a couple of issues while playing around (with a lot of users).  We need to implement the locking mechanism when not using redis (which was listed as a TODO a long time ago).

I'm doing a small refactor of the backend + typings to minimize the risk of errors, then it should be good to go

allardy committed:
chore(broadcast): small refactor of backend logic (#3356)

allardy committed:
Merge branch 'minor/12-10-0' into sm-broadcast

allardy committed:
fix date selection

allardy committed:
fix query for postgres

allardy committed:
hide unused field
